src/core_editor/editor.rs

```diff
@@ -176,60 +174,27 @@ impl Editor {
         self.line_buffer.move_to_line_end();
     }
 
-    fn get_index_undo(&self) -> usize {
-        if let Some(c) = self.edits.len().checked_sub(self.index_undo) {
-            c
-        } else {
-            0
-        }
-    }
-
     fn undo(&mut self) {
-        // NOTE: Try-blocks should help us get rid of this indirection too
-        self.undo_internal();
+        let val = self.edit_stack.undo();
+        self.line_buffer = val.clone();
     }
 
     fn redo(&mut self) {
-        // NOTE: Try-blocks should help us get rid of this indirection too
-        self.redo_internal();
-    }
-
-    fn redo_internal(&mut self) -> Option<()> {
-        if self.index_undo > 2 {
-            self.index_undo = self.index_undo.checked_sub(2)?;
-            self.undo_internal()
-        } else {
-            None
-        }
-    }
-
-    fn undo_internal(&mut self) -> Option<()> {
-        self.line_buffer = self.edits.get(self.get_index_undo())?.clone();
-
-        if self.index_undo <= self.edits.len() {
-            self.index_undo = self.index_undo.checked_add(1)?;
-        }
-        Some(())
+        let val = self.edit_stack.redo();
+        self.line_buffer = val.clone();
     }
 
     pub fn remember_undo_state(&mut self, is_after_action: bool) -> Option<()> {
-        self.reset_index_undo();
-
-        if self.edits.len() > 1
-            && self.edits.last()?.word_count() == self.line_buffer.word_count()
+        if self.edit_stack.current().word_count() == self.line_buffer.word_count()
             && !is_after_action
         {
-            self.edits.pop();
+            self.edit_stack.undo();
         }
-        self.edits.push(self.line_buffer.clone());
+        self.edit_stack.insert(self.line_buffer.clone());
 
         Some(())
     }
 
```

Comment:
Q: Should this operation be encapsulated as well? and just take the UndoBehavior enum for the op?

Reply:
OK, do you want to hold of @sherubthakur? I would be happy enough to merge

Reply:
Yup let's hold off that discussion (i.e. not make it part of this PR). I think this particular PR is mergable too.